packages/context/src/lib/controllers/context-provider.ts

```diff
@@ -94,13 +94,18 @@ export class ContextProvider<
     ev: ContextRequestEvent<Context<unknown, unknown>>
   ): void => {
     // Only call the callback if the context matches.
+    if (ev.context !== this.context) {
+      return;
+    }
     // Also, in case an element is a consumer AND a provider
     // of the same context, we want to avoid the element to self-register.
-    // The check on composedPath (as opposed to ev.target) is to cover cases
-    // where the consumer is in the shadowDom of the provider (in which case,
-    // event.target === this.host because of event retargeting).
+    if (ev.target === this.host) {
```

Comment:
I think this needs to be 
```suggestion
    if (ev.target !== this.host) {
```

It's only if the target and the host are the same that we need to disambiguate with the composed path check. If they're different, then we know the consumer and provider are not the same element.